### PR TITLE
Leverage JsonSchema and support complex JSON schema syntax such as $ref

### PR DESCRIPTION
**Pull request by vicancy**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Exceptions/SchemaKeywordNotSupported.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Microsoft.DocAsCode.Exceptions
+{
+    using System;
+    using System.Runtime.Serialization;
+
+    [Serializable]
+    public class SchemaKeywordNotSupported : DocfxException
+    {
+        public SchemaKeywordNotSupported() : this("This")
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/Exceptions/SchemaKeywordNotSupported.cs]: >This [](start = 51, length = 4)

should be a sentence?

Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Build.SchemaDriven/Exceptions/SchemaKeywordNotSupported.cs]: and rename class end with "Exception"

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs [hunk 2]
```diff
@@ -5,68 +5,54 @@ namespace Microsoft.DocAsCode.Build.SchemaDriven
...
-    using System.Threading;
 
     using Newtonsoft.Json;
-    using Newtonsoft.Json.Converters;
     using Newtonsoft.Json.Linq;
     using Newtonsoft.Json.Schema;
-    using Newtonsoft.Json.Serialization;
 
     using Microsoft.DocAsCode.Exceptions;
+    using System.Collections.Generic;
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs]: sort

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs [hunk 3]
```diff
@@ -5,68 +5,54 @@ namespace Microsoft.DocAsCode.Build.SchemaDriven
...
+        public JSchema InnerJSchema { get; private set; }
 
-        public JObject ToJObject()
-        {
-            return JObject.FromObject(this, DefaultSerializer.Value);
-        }
+        public JObject InnerJObject { get; private set; }
 
-        public static DocumentSchema Load(TextReader reader, string title)
+        public static DocumentSchema Load(StreamReader reader, string title)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs]: use `TextReader`?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs [hunk 4]
```diff
@@ -121,5 +107,95 @@ public static DocumentSchema Load(string schemaPath)
...
+                {
+                    throw new SchemaKeywordNotSupportedException(name);
+                }
+            }
+            else if (keyword is IDictionary<string, JSchema> dict)
+            {
+                if (dict.Count > 0)
+                {
+                    throw new SchemaKeywordNotSupportedException(name);
+
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs]: remove empty line

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs [hunk 5]
```diff
@@ -121,5 +107,95 @@ public static DocumentSchema Load(string schemaPath)
...
+        private static void CheckForNotSupportedKeyword(object keyword, string name)
+        {
+            if (keyword == null)
+            {
+                return;
+            }
+
+            if (keyword is IList<JSchema> list)
+            {
+                if (list.Count > 0)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs]: >> [](start = 31, length = 1)

should be == ?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs [hunk 6]
```diff
@@ -121,5 +107,95 @@ public static DocumentSchema Load(string schemaPath)
...
+                {
+                    throw new SchemaKeywordNotSupportedException(name);
+                }
+            }
+            else if (keyword is IDictionary<string, JSchema> dict)
+            {
+                if (dict.Count > 0)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs]: >> [](start = 31, length = 1)

should be == ?

File: src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs (hunk 5 above)
Comment (vicancy) [reply src/Microsoft.DocAsCode.Build.SchemaDriven/Models/DocumentSchema.cs]: `>0` means it contains value, however the feature is not supported.

**Pull request merged**